rust/main/chains/hyperlane-cosmos/src/providers/cosmos.rs

```diff
@@ -1,3 +1,328 @@
-pub use provider::CosmosProvider;
+use cosmrs::{
+    crypto::PublicKey,
+    tx::{SequenceNumber, SignerInfo},
+    AccountId, Coin, Tx,
+};
+use itertools::Itertools;
+use time::OffsetDateTime;
+use tonic::async_trait;
+use tracing::warn;
 
-mod provider;
+use hyperlane_core::{
+    h512_to_bytes, rpc_clients::BlockNumberGetter, utils::to_atto, BlockInfo,
+    ChainCommunicationError, ChainInfo, ChainResult, ContractLocator, HyperlaneChain,
+    HyperlaneDomain, HyperlaneProvider, HyperlaneProviderError, ReorgPeriod, TxnInfo,
+    TxnReceiptInfo, H256, H512, U256,
+};
+use hyperlane_metric::prometheus_metric::PrometheusClientMetrics;
+
+use crate::{utils, ConnectionConf, CosmosAccountId, CosmosAddress, HyperlaneCosmosError, Signer};
+
+use super::{grpc::GrpcProvider, rpc::RpcProvider};
+
+/// Trait for the QueryClient to be used in the CosmosProvider
+#[async_trait]
+pub trait BuildableQueryClient: Sized + std::fmt::Debug + Sync + Send + 'static + Clone {
+    /// Build the QueryClient
+    fn build_query_client(
+        grpc: GrpcProvider,
+        conf: &ConnectionConf,
+        locator: &ContractLocator,
+        signer: Option<Signer>,
+    ) -> ChainResult<Self>;
+
+    /// Whether or not the given address is a contract
+    async fn is_contract(&self, address: &H256) -> ChainResult<bool>;
+
+    /// Extract the message recipient contract address from the tx
+    /// this is implementation specific
+    fn parse_tx_message_recipient(&self, tx: &Tx, hash: &H512) -> ChainResult<H256>;
+}
+
+/// Cosmos Native Provider
```

Comment:
I suppose `CosmosProvider<QueryClient>` is a provider which is used in both CosmWasm and Cosmos Native.

Reply:
Correct! The generic part will take core of the implementation specific queries - this way they can share a large amount of logic

Reply:
This is just a note on the comment.

```suggestion
/// Cosmos Provider
```